File: py/day7-1.py
from dataclasses import dataclass
from typing import Optional

FILE_SIZE_LIMIT = 100000


@dataclass
class File:
    size: int


@dataclass
class Folder:
    files: list[File]
    folders: dict[str, "Folder"]
    parent: Optional["Folder"]

    def get_sizes(self, contents: list) -> int:
        self_files = sum(file.size for file in self.files)
        self_folders = sum(
            folder.get_sizes(contents) for folder in self.folders.values()
        )
        contents.append(self_files + self_folders)
        return self_files + self_folders


def main() -> None:
    cd = Folder([], {}, None)
    all_folders: list[int] = []

    while True:
        tokens = input().split()
        match tokens:
            case ["$", "cd", "/"]:
                while cd.parent is not None:
                    cd = cd.parent
            case ["$", "cd", ".."]:
                cd = cd.parent or cd
            case ["$", "cd", name]:
                if name in cd.folders:
                    cd = cd.folders[name]
                else:
                    cd.folders[name] = Folder([], {}, cd)
                    cd = cd.folders[name]
            case ["$", *_]:
                pass
            case ["dir", name]:
                if name not in cd.folders:
                    cd.folders[name] = Folder([], {}, cd)
            case [size, _]:
                cd.files.append(File(int(size)))
            case ["sum"]:
                while cd.parent is not None:
                    cd = cd.parent

                cd.get_sizes(all_folders)
                break
            case _:
                pass

    print("Criterion:", sum(filter(lambda x: x <= FILE_SIZE_LIMIT, all_folders)))


if __name__ == "__main__":
    main()

File: src/day7-1.rs
// doesn't work :(
// linked lists in rust are painful
// especially coming from python

use std::{collections::HashMap, io::stdin};

#[derive(Debug, Clone)]
struct File {
    size: u32,
}

#[derive(Debug, Clone)]
struct Folder<'a> {
    files: Vec<File>,
    folders: HashMap<String, &'a Folder<'a>>,
    parent: Option<&'a mut Folder<'a>>,
}

impl Folder<'_> {
    pub fn get_size(&self) -> u32 {
        let folder_total: u32 = self.folders.iter().map(|(_, f)| f.get_size()).sum();
        let files_total: u32 = self.files.iter().map(|f| f.size).sum();

        files_total + folder_total
    }
}

fn main() {
    let mut s = String::new();
    let mut cd = &mut Folder {
        files: vec![],
        folders: HashMap::new(),
        parent: None,
    };

    loop {
        s.clear();
        stdin().read_line(&mut s).unwrap();

        let tokens: Vec<&str> = s.trim().split(' ').collect();

        let &first = tokens.first().expect("First is empty!");
        let &second = tokens.get(1).expect("No commmand!");
        let third = tokens.get(2);
        if first == "$" {
            // new command
            if second == "cd" {
                match third {
                    Some(&"..") => {
                        let mut new_cd = cd.parent.clone();
                        cd = new_cd.expect("No parent!");
                    }
                    Some(&path) => {
                        cd = {
                            if let Some(child) = cd.folders.get_mut(path) {
                                child
                            } else {
                                let new = &mut Folder {
                                    files: vec![],
                                    folders: HashMap::new(),
                                    parent: Some(cd),
                                };
                                cd.folders.insert(path.to_string(), new);
                                new
                            }
                        };
                    }
                    _ => panic!("bad things happened"),
                }
            } else if second == "ls" {
            }
        }
    }
}
